Source/WebCore/style/AnchorPositionEvaluator.cpp

```diff
@@ -741,9 +741,59 @@ static bool firstChildPrecedesSecondChild(const RenderObject* firstChild, const
     return false;
 }
 
+// Given an anchor element and its anchor names, locate the closest ancestor element
+// that establishes an anchor scope affecting this anchor element, and return the pointer
+// to such element. If no ancestor establishes an anchor scope affecting this anchor,
+// returns nullptr.
+static CheckedPtr<Element> anchorScopeForAnchorElement(const Element& anchorElement, const Vector<ScopedName>& anchorNames)
+{
+    // Precondition: anchorElement is an anchor, which has at least one anchor name.
+    ASSERT(!anchorNames.isEmpty());
+
+    // Traverse up the element tree through each ancestor.
+    for (CheckedPtr currentAncestor = anchorElement.parentElement(); currentAncestor; currentAncestor = currentAncestor->parentElement()) {
+        CheckedPtr currentAncestorStyle = currentAncestor->renderStyle();
+        if (!currentAncestorStyle)
+            continue;
+
+        const auto& currentAncestorAnchorScope = currentAncestorStyle->anchorScope();
+        switch (currentAncestorAnchorScope.type) {
+        // Does not establish a scope.
+        case NameScope::Type::None:
+            continue;
+
+        // Scopes all anchors that are descendants of the current ancestor.
+        case NameScope::Type::All:
+            return currentAncestor;
+
+        // Scopes anchors that are (1) descendants of the current ancestor and
+        // (2) its name is specified in the scope.
+        case NameScope::Type::Ident:
+            for (const auto& anchorName : anchorNames) {
+                if (currentAncestorAnchorScope.names.contains(anchorName.name))
+                    return currentAncestor;
+            }
+
+            continue;
+        }
+    }
+
+    return nullptr;
+}
+
 // See: https://drafts.csswg.org/css-anchor-position-1/#acceptable-anchor-element
 static bool isAcceptableAnchorElement(const RenderBoxModelObject& anchorRenderer, Ref<const Element> anchorPositionedElement)
 {
+    // "Possible anchor is either an element or a fully styleable tree-abiding pseudo-element."
+    // This always have an associated Element (for ::before/::after it is PseudoElement).
+    if (!anchorRenderer.element())
+        return false;
+
+    if (auto anchorScopeElement = anchorScopeForAnchorElement(*anchorRenderer.element(), anchorRenderer.style().anchorNames())) {
+        if (!anchorPositionedElement->isDescendantOf(*anchorScopeElement))
+            return false;
```

Comment:
Similarly this should probably operate in the flat tree. I don't think there `isDescendantInComposedTree` helper currently but you can just iterate with parentElementInComposedTree.